homeassistant/components/modbus/__init__.py

```diff
@@ -178,73 +146,136 @@ def write_coil(service):
         client_name = service.data[ATTR_HUB]
         hub_collect[client_name].write_coil(unit, address, state)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_modbus)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_modbus)
 
     return True
 
 
 class ModbusHub:
-    """Thread safe wrapper class for pymodbus."""
+    """Wrapper class for pymodbus."""
 
-    def __init__(self, modbus_client, name):
+    def __init__(self, client_config, main_loop):
         """Initialize the Modbus hub."""
-        self._client = modbus_client
-        self._lock = threading.Lock()
-        self._name = name
+        self._client = None
+        _LOGGER.debug("Preparing setup: %s", client_config)
+
+        # generic configuration
+        self._loop = main_loop
+        self._config_name = client_config[CONF_NAME]
+        self._config_type = client_config[CONF_TYPE]
+        self._config_port = client_config[CONF_PORT]
+        self._config_timeout = client_config[CONF_TIMEOUT]
+        self._config_delay = client_config[CONF_DELAY]
+
+        if self._config_type == "serial":
+            # serial configuration
+            self._config_method = client_config[CONF_METHOD]
+            self._config_baudrate = client_config[CONF_BAUDRATE]
+            self._config_stopbits = client_config[CONF_STOPBITS]
+            self._config_bytesize = client_config[CONF_BYTESIZE]
+            self._config_parity = client_config[CONF_PARITY]
+        else:
+            # network configuration
+            self._config_host = client_config[CONF_HOST]
 
     @property
     def name(self):
         """Return the name of this hub."""
-        return self._name
+        return self._config_name
+
+    async def _connect_delay(self):
+        if self._config_delay > 0:
+            await asyncio.sleep(self._config_delay)
+            self._config_delay = 0
+
+    def setup(self):
```

Comment:
It doesn't look like there is any I/O here so you can mark this async safe with the @callback decorator
```
from homeassistant.core import callback
```
```suggestion
    @callback
    def async_setup(self):
```

Reply:
There are actually IO involved the client* connect to the target. However for some (to me) unknown reason the client* cannot be called from an async method.